lib/tree.rb

```diff
@@ -16,51 +16,114 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
+  end
+
+  def add_helper(root, key, value)
+    if root.nil?
+      root = TreeNode.new(key, value)
+      return root
+    end
+
+    if root.key > key
+      root.left = add_helper(root.left, key, value)
+    else
+      root.right = add_helper(root.right, key, value)
+    end
+
+    root
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def find(key)
```

Comment:
👍 The time complexity is O(log n) if the tree is balanced and O(n) if it's unbalanced.

The space complexity is the same (because of the recursion).